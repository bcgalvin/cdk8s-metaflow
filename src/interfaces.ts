import { HttpIngressPathType, ImagePullPolicy, ServiceType } from './enums';
import { Affinity, EnvFromSource, IngressTls, SecurityContext, Toleration } from './imports/k8s';

export interface ImageOptions {
  readonly repository?: string;
  readonly pullPolicy?: ImagePullPolicy;
  readonly tag?: string;
}

export interface ServiceAccountOptions {
  readonly create?: boolean | undefined;
  readonly annotations?: Record<string, string>;
  readonly name?: string;
}

export interface DbMigrationOptions {
  readonly runOnStart?: boolean;
  readonly onlyIfDbEmpty?: boolean;
}

export interface ServiceOptions {
  readonly type?: ServiceType;
  readonly port?: number;
  readonly annotations?: Record<string, string>;
}

export interface HttpIngressPath {
  readonly path?: string;
  readonly pathType?: HttpIngressPathType;
}

export interface HostOptions {
  readonly host?: string;
  readonly paths?: HttpIngressPath[];
}

export interface IngressOptions {
  readonly enabled?: boolean;
  readonly className?: string;
  readonly annotations?: Record<string, string>;
  readonly hosts?: HostOptions[];
  readonly tls?: IngressTls;
}

export interface AutoscalingOptions {
  readonly enabled?: boolean;
  readonly minReplicas?: number;
  readonly maxReplicas?: number;
  readonly targetCPUUtilizationPercentage?: number;
}

export interface IApiResource {
  readonly apiGroup?: string;
  readonly resourceType?: string;
  readonly resourceName?: string;
}

export interface MetadataDatabaseEnvOptions {
  readonly port?: number;
  readonly password?: string;
  readonly host?: string;
  readonly user?: string;
}

export interface MetaflowServiceOptions {
  readonly replicaCount?: number;
  readonly image?: ImageOptions;
  readonly imagePullSecrets?: string[];
  readonly nameOverride?: string;
  readonly fullnameOverride?: string;
  readonly serviceAccount?: ServiceAccountOptions;
  readonly podAnnotations?: Record<string, string>;
  readonly podSecurityContext?: Record<string, string>;
  readonly dbMigrations?: DbMigrationOptions;
  readonly securityContext?: SecurityContext;
  readonly service?: ServiceOptions;
  readonly ingress?: IngressOptions;
  readonly resources?: IApiResource[];
  readonly autoscaling?: AutoscalingOptions;
  readonly nodeSelector?: Record<string, string>;
  readonly tolerations?: Toleration[];
  readonly affinity?: Affinity;
  readonly envFrom?: EnvFromSource[];
  readonly metadatadb?: MetadataDatabaseEnvOptions;
}

export interface MetaflowUIOptions {
  readonly replicaCount?: number;
  readonly image?: ImageOptions;
  readonly uiImage?: ImageOptions;
  readonly imagePullSecrets?: string[];
  readonly nameOverride?: string;
  readonly fullnameOverride?: string;
  readonly env: Record<string, string>;
  readonly serviceAccount?: ServiceAccountOptions;
  readonly podAnnotations?: Record<string, string>;
  readonly podSecurityContext?: Record<string, string>;
  readonly securityContext?: SecurityContext;
  readonly service?: ServiceOptions;
  readonly serviceStatic?: ServiceOptions;
  readonly ingress?: IngressOptions;
  readonly resources?: IApiResource[];
  readonly autoscaling?: AutoscalingOptions;
  readonly nodeSelector?: Record<string, string>;
  readonly tolerations?: Toleration[];
  readonly affinity?: Affinity;
  readonly envFrom?: EnvFromSource[];
  readonly metadatadb?: MetadataDatabaseEnvOptions;
  readonly metaflowDatastoreSysrootS3?: string;
}

export interface DatabaseAuthOptions {
  readonly database?: string;
  readonly password?: string;
  readonly username?: string;
  readonly postgresPassword?: string;
  readonly enablePostgresUser?: boolean;
  readonly replicationPassword?: string;
  readonly replicationUsername?: string;
}

export interface DatabaseMetricsOptions {
  readonly enabled?: boolean;
}

export interface DatabaseResourceRequestOptions {
  readonly memory?: string;
  readonly cpu?: string;
}

export interface DatabaseResourcesOptions {
  readonly requests?: DatabaseResourceRequestOptions;
}

export interface DatabaseReplicationOptions {
  readonly enabled?: boolean;
  readonly readReplicas?: number;
}

export interface DatabaseVolumePermissionsOptions {
  readonly enabled?: boolean;
}

export interface MetadataDatabaseOptions {
  readonly auth?: DatabaseAuthOptions;
  readonly metrics?: DatabaseMetricsOptions;
  readonly resources?: DatabaseResourcesOptions;
  readonly replication?: DatabaseReplicationOptions;
  readonly architecture?: string;
  readonly volumePermissions?: DatabaseVolumePermissionsOptions;
}
