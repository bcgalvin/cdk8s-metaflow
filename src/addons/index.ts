export * from './addon';
export * from './postgres';
export * from './argo-workflows';
export * from './autoscaler';
