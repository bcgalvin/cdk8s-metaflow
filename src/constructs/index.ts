export * from './metaflow-service';
export * from './interfaces';
