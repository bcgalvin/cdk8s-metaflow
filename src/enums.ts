export enum ImagePullPolicy {
  ALWAYS = 'Always',
  IF_NOT_PRESENT = 'IfNotPresent',
  NEVER = 'Never',
}

export enum HttpIngressPathType {
  PREFIX = 'Prefix',
  EXACT = 'Exact',
  IMPLEMENTATION_SPECIFIC = 'ImplementationSpecific',
}

export enum ServiceType {
  CLUSTER_IP = 'ClusterIP',
  NODE_PORT = 'NodePort',
  LOAD_BALANCER = 'LoadBalancer',
  EXTERNAL_NAME = 'ExternalName',
}
