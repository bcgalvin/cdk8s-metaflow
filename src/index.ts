export * from './metaflow-service';
export * as k8s from './imports/k8s';
export * as minio from './imports/minio.min.io';
