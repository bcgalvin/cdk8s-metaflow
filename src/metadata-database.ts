import { Helm } from 'cdk8s';
import { Construct } from 'constructs';
import { MetadataDatabaseOptions } from './interfaces';

export interface MetadataDatabaseProps {
  readonly chartVersion: string;
  readonly chartValues?: MetadataDatabaseOptions;
}

export class MetadataDatabase extends Construct {
  constructor(scope: Construct, id: string, props: MetadataDatabaseProps) {
    super(scope, id);

    const chartValues = props.chartValues ?? {
      enabled: true,
      auth: {
        database: 'metaflow',
        password: 'metaflow',
        username: 'metaflow',
      },
    };

    new Helm(this, 'postgres', {
      chart: 'bitnami/postgresql',
      releaseName: 'release-name-postgresql',
      helmFlags: ['--version', props.chartVersion],
      values: chartValues,
    });
  }
}
