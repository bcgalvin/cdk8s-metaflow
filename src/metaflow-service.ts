import { Helm } from 'cdk8s';
import * as kplus from 'cdk8s-plus-22';
import { HttpIngressPathType, Protocol } from 'cdk8s-plus-22';
import { Construct } from 'constructs';

export interface PostgresOptions {
  readonly databaseName?: string;
  readonly databaseUser?: string;
  readonly databasePassword?: string;
}

export interface IngressOptions {
  readonly hostName?: string;
}

export interface MetaflowServiceProps {
  readonly serviceAccountName?: string;
  readonly serviceName?: string;
  readonly image?: string;
  readonly metadataServicePort?: number;
  readonly upgradesServicePort?: number;
  readonly postgresEnabled?: boolean;
  readonly postgresOptions?: PostgresOptions;
  readonly ingressEnabled?: boolean;
  readonly ingressOptions?: IngressOptions;
}

export class MetaflowService extends Construct {
  public readonly deployment: kplus.Deployment;
  public readonly service: kplus.Service;

  constructor(scope: Construct, id: string, props?: MetaflowServiceProps) {
    super(scope, id);

    const serviceAccountName = props?.serviceAccountName || 'release-name-metaflow-service';
    const serviceName = props?.serviceName || 'metaflow-service';
    const image = props?.image || 'public.ecr.aws/outerbounds/metaflow_metadata_service:2.2.4';
    const metadataServicePort = props?.metadataServicePort || 8080;
    const upgradesServicePort = props?.upgradesServicePort || 8082;
    const ingressHostName = props?.ingressOptions?.hostName || 'metaflow-service.local';

    const dbEnvVals: { [p: string]: kplus.EnvValue } = {
      MF_METADATA_DB_NAME: kplus.EnvValue.fromValue(props?.postgresOptions?.databaseName || 'metaflow'),
      MF_METADATA_DB_USER: kplus.EnvValue.fromValue(props?.postgresOptions?.databaseUser || 'metaflow'),
      MF_METADATA_DB_PSWD: kplus.EnvValue.fromValue(props?.postgresOptions?.databasePassword || 'metaflow'),
      MF_METADATA_DB_PORT: kplus.EnvValue.fromValue('5432'),
      MF_METADATA_DB_HOST: kplus.EnvValue.fromValue('release-name-postgresql'),
    };
    const serviceAccount = new kplus.ServiceAccount(this, 'service-account', {
      metadata: {
        name: serviceAccountName,
        labels: {
          'helm.sh/chart': 'metaflow-service-0.2.0',
          'app.kubernetes.io/name': serviceName,
          'app.kubernetes.io/instance': 'release-name',
          'app.kubernetes.io/version': '2.2.4',
          'app.kubernetes.io/managed-by': 'Helm',
        },
      },
    });

    this.deployment = new kplus.Deployment(this, 'metaflow-deployment', {
      replicas: 1,
      initContainers: [
        {
          name: 'db-migrations',
          image: image,
          command: ['/opt/latest/bin/python3', '/root/run_goose.py', '--only-if-empty-db'],
          envVariables: dbEnvVals,
        },
      ],
      containers: [
        {
          name: 'metaflow-service',
          image: image,
          command: ['/opt/latest/bin/python3', '-m', 'services.metadata_service.server'],
          imagePullPolicy: kplus.ImagePullPolicy.IF_NOT_PRESENT,
          liveness: kplus.Probe.fromHttpGet('/ping', { port: metadataServicePort }),
          readiness: kplus.Probe.fromHttpGet('/ping', { port: metadataServicePort }),
          port: metadataServicePort,
          envVariables: dbEnvVals,
        },
      ],
      metadata: {
        name: serviceAccountName,
        labels: {
          'helm.sh/chart': 'metaflow-service-0.2.0',
          'app.kubernetes.io/name': serviceName,
          'app.kubernetes.io/instance': 'release-name',
          'app.kubernetes.io/version': '2.2.4',
          'app.kubernetes.io/managed-by': 'Helm',
        },
      },
      serviceAccount,
    });

    this.service = new kplus.Service(this, 'metaflow-service', {
      metadata: {
        name: serviceAccountName,
        labels: {
          'helm.sh/chart': 'metaflow-service-0.2.0',
          'app.kubernetes.io/name': serviceName,
          'app.kubernetes.io/instance': 'release-name',
          'app.kubernetes.io/version': '2.2.4',
          'app.kubernetes.io/managed-by': 'Helm',
        },
      },
      type: kplus.ServiceType.CLUSTER_IP,
      ports: [
        {
          name: 'metadata',
          port: metadataServicePort,
          protocol: Protocol.TCP,
          targetPort: metadataServicePort,
        },
        {
          name: 'upgrades',
          port: upgradesServicePort,
          protocol: Protocol.TCP,
          targetPort: upgradesServicePort,
        },
      ],
    });

    new kplus.Pod(this, 'metaflow-service-pod', {
      metadata: {
        name: 'release-name-metaflow-service-test-connection',
        labels: {
          'helm.sh/chart': 'metaflow-service-0.2.0',
          'app.kubernetes.io/name': serviceName,
          'app.kubernetes.io/instance': 'release-name',
          'app.kubernetes.io/version': '2.2.4',
          'app.kubernetes.io/managed-by': 'Helm',
        },
        annotations: { 'helm.sh/hook': 'test' },
      },
      containers: [
        {
          name: 'wget',
          image: 'busybox',
          command: ['wget'],
          args: ['release-name-metaflow-service:8080'],
        },
      ],
      restartPolicy: kplus.RestartPolicy.ON_FAILURE,
    });

    if (props?.postgresEnabled) {
      new Helm(this, 'postgres', {
        chart: 'bitnami/postgresql',
        releaseName: 'release-name-postgresql',
        values: {
          enabled: true,
          serviceAccount: {
            name: serviceAccountName,
          },
          auth: {
            database: props?.postgresOptions?.databaseName ?? 'metaflow',
            username: props?.postgresOptions?.databaseUser ?? 'metaflow',
            password: props?.postgresOptions?.databasePassword ?? 'metaflow',
          },
        },
      });
    }

    if (props?.ingressEnabled) {
      const ingress = new kplus.Ingress(this, 'ingress', {
        metadata: {
          name: serviceAccountName,
          labels: {
            'helm.sh/chart': 'metaflow-service-0.2.0',
            'app.kubernetes.io/name': serviceName,
            'app.kubernetes.io/instance': 'release-name',
            'app.kubernetes.io/version': '2.2.4',
            'app.kubernetes.io/managed-by': 'Helm',
          },
        },
      });
      ingress.addHostRule(
        ingressHostName,
        '/',
        kplus.IngressBackend.fromService(this.service, { port: metadataServicePort }),
        HttpIngressPathType.IMPLEMENTATION_SPECIFIC,
      );
    }
  }
}
