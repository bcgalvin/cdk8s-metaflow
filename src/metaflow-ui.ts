import { Helm } from 'cdk8s';
import { Construct } from 'constructs';

export interface MetaflowUIProps {
  readonly chartVersion: string;
}

export class MetaflowUI extends Construct {
  constructor(scope: Construct, id: string, props: MetaflowUIProps) {
    super(scope, id);

    new Helm(this, 'metaflow-ui-helm', {
      chart: 'metaflow/metaflow-ui',
      releaseName: 'release-name-metaflow-ui',
      helmFlags: ['--version', props.chartVersion],
    });
  }
}
