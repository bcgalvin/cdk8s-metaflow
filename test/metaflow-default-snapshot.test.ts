import { MetaflowService } from '../src';
// @ts-ignore
import { synth } from './util';

describe('metaflow service', () => {
  test('default snapshot', () => {
    // GIVEN
    const chart = synth(MetaflowService);
    // THEN
    expect(chart).toMatchSnapshot();
  });
});
