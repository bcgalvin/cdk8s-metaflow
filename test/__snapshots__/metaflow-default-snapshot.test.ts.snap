// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metaflow service default snapshot 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "automountServiceAccountToken": true,
    "kind": "ServiceAccount",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-service",
        "app.kubernetes.io/version": "2.2.4",
        "helm.sh/chart": "metaflow-service-0.2.0",
      },
      "name": "release-name-metaflow-service",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-service",
        "app.kubernetes.io/version": "2.2.4",
        "helm.sh/chart": "metaflow-service-0.2.0",
      },
      "name": "release-name-metaflow-service",
    },
    "spec": Object {
      "minReadySeconds": 0,
      "progressDeadlineSeconds": 600,
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.io/metadata.addr": "construct-metaflow-deployment-c89bf212",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": "25%",
          "maxUnavailable": "25%",
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.io/metadata.addr": "construct-metaflow-deployment-c89bf212",
          },
        },
        "spec": Object {
          "automountServiceAccountToken": true,
          "containers": Array [
            Object {
              "command": Array [
                "/opt/latest/bin/python3",
                "-m",
                "services.metadata_service.server",
              ],
              "env": Array [
                Object {
                  "name": "MF_METADATA_DB_NAME",
                  "value": "metaflow",
                },
                Object {
                  "name": "MF_METADATA_DB_USER",
                  "value": "metaflow",
                },
                Object {
                  "name": "MF_METADATA_DB_PSWD",
                  "value": "metaflow",
                },
                Object {
                  "name": "MF_METADATA_DB_PORT",
                  "value": "5432",
                },
                Object {
                  "name": "MF_METADATA_DB_HOST",
                  "value": "release-name-postgresql",
                },
              ],
              "image": "public.ecr.aws/outerbounds/metaflow_metadata_service:2.2.4",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "failureThreshold": 3,
                "httpGet": Object {
                  "path": "/ping",
                  "port": 8080,
                },
              },
              "name": "metaflow-service",
              "ports": Array [
                Object {
                  "containerPort": 8080,
                },
              ],
              "readinessProbe": Object {
                "failureThreshold": 3,
                "httpGet": Object {
                  "path": "/ping",
                  "port": 8080,
                },
              },
              "securityContext": Object {
                "privileged": false,
                "readOnlyRootFilesystem": false,
                "runAsNonRoot": false,
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "initContainers": Array [
            Object {
              "command": Array [
                "/opt/latest/bin/python3",
                "/root/run_goose.py",
                "--only-if-empty-db",
              ],
              "env": Array [
                Object {
                  "name": "MF_METADATA_DB_NAME",
                  "value": "metaflow",
                },
                Object {
                  "name": "MF_METADATA_DB_USER",
                  "value": "metaflow",
                },
                Object {
                  "name": "MF_METADATA_DB_PSWD",
                  "value": "metaflow",
                },
                Object {
                  "name": "MF_METADATA_DB_PORT",
                  "value": "5432",
                },
                Object {
                  "name": "MF_METADATA_DB_HOST",
                  "value": "release-name-postgresql",
                },
              ],
              "image": "public.ecr.aws/outerbounds/metaflow_metadata_service:2.2.4",
              "imagePullPolicy": "Always",
              "name": "db-migrations",
              "securityContext": Object {
                "privileged": false,
                "readOnlyRootFilesystem": false,
                "runAsNonRoot": false,
              },
            },
          ],
          "securityContext": Object {
            "fsGroupChangePolicy": "Always",
            "runAsNonRoot": false,
          },
          "serviceAccountName": "release-name-metaflow-service",
          "setHostnameAsFQDN": false,
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-service",
        "app.kubernetes.io/version": "2.2.4",
        "helm.sh/chart": "metaflow-service-0.2.0",
      },
      "name": "release-name-metaflow-service",
    },
    "spec": Object {
      "externalIPs": Array [],
      "ports": Array [
        Object {
          "name": "metadata",
          "port": 8080,
          "protocol": "TCP",
          "targetPort": 8080,
        },
        Object {
          "name": "upgrades",
          "port": 8082,
          "protocol": "TCP",
          "targetPort": 8082,
        },
      ],
      "selector": Object {},
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": Object {
      "annotations": Object {
        "helm.sh/hook": "test",
      },
      "labels": Object {
        "app.kubernetes.io/instance": "release-name",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "metaflow-service",
        "app.kubernetes.io/version": "2.2.4",
        "cdk8s.io/metadata.addr": "construct-metaflow-service-pod-c8290cca",
        "helm.sh/chart": "metaflow-service-0.2.0",
      },
      "name": "release-name-metaflow-service-test-connection",
    },
    "spec": Object {
      "automountServiceAccountToken": true,
      "containers": Array [
        Object {
          "args": Array [
            "release-name-metaflow-service:8080",
          ],
          "command": Array [
            "wget",
          ],
          "image": "busybox",
          "imagePullPolicy": "Always",
          "name": "wget",
          "securityContext": Object {
            "privileged": false,
            "readOnlyRootFilesystem": false,
            "runAsNonRoot": false,
          },
        },
      ],
      "dnsPolicy": "ClusterFirst",
      "restartPolicy": "OnFailure",
      "securityContext": Object {
        "fsGroupChangePolicy": "Always",
        "runAsNonRoot": false,
      },
      "setHostnameAsFQDN": false,
    },
  },
]
`;
